### PR TITLE
arch/risc-v/common: provide architecture specific perfmon bindings.

### PR DESCRIPTION
**Pull request by g2gps**
## Summary
provide architecture specific perfmon bindings.

Using the up_perf_xx bindings directory with the core clock is more efficient than performing a nanosecond conversion on every gettime event.

## Impact

A nanosecond conversion is no longer required each time a perf_gettime is used.

## Testing

Performance monitoring subsystem is used on custom hardware for testing. The arty_a7:knsh-tickless configuration builds, however the subsystem isn't used.

### CODE REVIEW

File: arch/risc-v/src/litex/litex_perf.c [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+void up_perf_init(void *arg)
+{
+  (void)arg;
+}
+
+/****************************************************************************
+ * Name: up_perf_gettime
+ ****************************************************************************/
+
+unsigned long up_perf_gettime(void)
```
Comment (xiaoxiang781216) [inline arch/risc-v/src/litex/litex_perf.c:59]: but it's a general implementation, could be moved to the common place

File: arch/risc-v/src/litex/litex_perf.c [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: up_perf_init
+ ****************************************************************************/
+
+void up_perf_init(void *arg)
+{
+  (void)arg;
+}
```
Comment (acassis) [inline arch/risc-v/src/litex/litex_perf.c:53]: Please include a description of this function and its goal

File: arch/risc-v/src/litex/litex_perf.c [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+
+/****************************************************************************
+ * Name: up_perf_gettime
+ ****************************************************************************/
+
+unsigned long up_perf_gettime(void)
+{
+  return READ_CSR(CSR_TIME);
```
[the remark this answers is not in the document]
Comment (g2gps) [reply arch/risc-v/src/litex/litex_perf.c:61]: I'm curious about this comment. Both CSR_CYCLE and CSR_TIME are part of the risc-v core. Is the a reason CSR_CYCLE is faster?

File: arch/risc-v/src/litex/litex_perf.c (hunk 1 above)
Comment (g2gps) [reply arch/risc-v/src/litex/litex_perf.c:59]: You're right, I'll move this to the common section and update as needed.

File: arch/risc-v/src/litex/litex_perf.c (hunk 2 above)
Comment (g2gps) [reply arch/risc-v/src/litex/litex_perf.c:53]: No problem, I'll update as part of the next patch set.

**Comment by acassis** [comment]
ping @g2gps

**Comment by g2gps** [comment]
> ping @g2gps

I was hoping for some feedback from @xiaoxiang781216 in regards to my comment above, before continuing with the PR.

**Comment by xiaoxiang781216** [comment]
> > ping @g2gps
> 
> I was hoping for some feedback from @xiaoxiang781216 in regards to my comment above, before continuing with the PR.

Yes, it's better to move to the common place.

**Comment by g2gps** [comment]
> > > ping @g2gps
> > 
> > 
> > I was hoping for some feedback from @xiaoxiang781216 in regards to my comment above, before continuing with the PR.
> 
> Yes, it's better to move to the common place.

Sorry, I was referring to my other comment, where you mentioned using `CSR_CYCLE`

File: arch/risc-v/src/litex/litex_perf.c [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+ * Public Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: up_perf_init
+ ****************************************************************************/
+
+void up_perf_init(void *arg)
+{
+  (void)arg;
```
Comment (pkarashchenko) [inline arch/risc-v/src/litex/litex_perf.c:52]: I think `UNUSED(arg)` could be used

**Comment by xiaoxiang781216** [comment]
> > > > ping @g2gps
> > > 
> > > 
> > > I was hoping for some feedback from @xiaoxiang781216 in regards to my comment above, before continuing with the PR.
> > 
> > 
> > Yes, it's better to move to the common place.
> 
> Sorry, I was referring to my other comment, where you mentioned using `CSR_CYCLE`

Done.

**Comment by g2gps** [comment]
A couple of questions have come up in the details whilst revisiting this:

- The `vexriscv_smp` core doesn't implement the performance counter by default. Opensbi handles this case on startup by attempting a read and trapping the exception.
- The currently implementation could be moved to `riscv_perf.c` instead, however some risc-v ports, e.g  `esp32c3_perf` already have there own architecture specific implementation, so there would need to be some conditional logic around this.
- Assuming the Litex architecture specific perf_xx implementation is kept, I'll need to update this to handle the standard `vexriscv` core, which only has the m-mode timer.

**Comment by g2gps** [comment]
> Let chip developer add riscv_perf.c to their Make.defs if they want to use the common implementation.

Yes, that would work for the general implementation.

However, this won't work for the vexriscv core. I'm not sure what the better approach is here. We could either:
 - Have a kconfig option which specifies if the performance monitor CSRs are implemented. Both the time based and counter based implementations are in the common location.
 - Just have the counter based implementation in the generic location. Architectures can provide their own implementation if needed.

The second option seems a bit cleaner, but I'd appreciate any suggestions.

**Comment by xiaoxiang781216** [comment]
> > Let chip developer add riscv_perf.c to their Make.defs if they want to use the common implementation.
> 
> Yes, that would work for the general implementation.
> 
> However, this won't work for the vexriscv core. I'm not sure what the better approach is here. We could either:
> 
> * Have a kconfig option which specifies if the performance monitor CSRs are implemented. Both the time based and counter based implementations are in the common location.
> * Just have the counter based implementation in the generic location. Architectures can provide their own implementation if needed.
> 
> The second option seems a bit cleaner, but I'd appreciate any suggestions.

How about we provide riscv_perf_cycle.c and riscv_perf-time.c? Let's chip developer add one of these as needed.

**Comment by g2gps** [comment]
> How about we provide riscv_perf_cycle.c and riscv_perf_time.c? Let's chip developer add one of these as needed.

This makes sense. I've updated as suggested. I've testing the CSR_TIME implementation on Litex. I believe I can test the cycle based implementation on QEMU, if I do a quick integration.

**Review by xiaoxiang781216 — APPROVED** [review]

**Comment by g2gps** [comment]
I did a quick test with `rv-virt:nsh`, by adding ` riscv_perf_cycle.c` and enabling `CONFIG_SCHED_IRQMONITOR`, `CONFIG_ARCH_PERF_EVENTS`

```
nsh> cat /proc/irqs
IRQ HANDLER  ARGUMENT    COUNT    RATE    TIME
 11 80005f16 00000000          2    4.385 10927
 23 80005856 8002ca28        405  888.157 41596
 37 80002d14 80028f88          2    4.385 58285
```

The calculated time is not correct, as I've used the `MTIMER_FREQ`. I don't know how to find the cycle frequency in qemu-rv

**Comment by xiaoxiang781216** [comment]
need fix this ci error:
```
  Normalize arty_a7/knsh-tickless
67d66
< CONFIG_PERF_OVERFLOW_CORRECTION=y
```

**Comment by g2gps** [comment]
> need fix this ci error:
> 
> ```
>   Normalize arty_a7/knsh-tickless
> 67d66
> < CONFIG_PERF_OVERFLOW_CORRECTION=y
> ```

Thanks, SYSTEM_TIME64 should have been enabled here as well. Since this is the width of the underlying system timer.

**Pull request merged**